## Basic ML Model using SkLearn Library
## Using Cancer Dataset

### Harjyot Singh Bagga - 18BCE0694

In [1]:
import sklearn.datasets
import numpy as np

In [2]:
# Loading the dataset
cancer_ds = sklearn.datasets.load_breast_cancer()

In [3]:
# differentiating the data features and the data target
X = cancer_ds.data
Y = cancer_ds.target

In [4]:
print(X)
print(Y)

[[1.799e+01 1.038e+01 1.228e+02 ... 2.654e-01 4.601e-01 1.189e-01]
 [2.057e+01 1.777e+01 1.329e+02 ... 1.860e-01 2.750e-01 8.902e-02]
 [1.969e+01 2.125e+01 1.300e+02 ... 2.430e-01 3.613e-01 8.758e-02]
 ...
 [1.660e+01 2.808e+01 1.083e+02 ... 1.418e-01 2.218e-01 7.820e-02]
 [2.060e+01 2.933e+01 1.401e+02 ... 2.650e-01 4.087e-01 1.240e-01]
 [7.760e+00 2.454e+01 4.792e+01 ... 0.000e+00 2.871e-01 7.039e-02]]
[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 1 0 0 0 0 0 0 0 0 1 0 1 1 1 1 1 0 0 1 0 0 1 1 1 1 0 1 0 0 1 1 1 1 0 1 0 0
 1 0 1 0 0 1 1 1 0 0 1 0 0 0 1 1 1 0 1 1 0 0 1 1 1 0 0 1 1 1 1 0 1 1 0 1 1
 1 1 1 1 1 1 0 0 0 1 0 0 1 1 1 0 0 1 0 1 0 0 1 0 0 1 1 0 1 1 0 1 1 1 1 0 1
 1 1 1 1 1 1 1 1 0 1 1 1 1 0 0 1 0 1 1 0 0 1 1 0 0 1 1 1 1 0 1 1 0 0 0 1 0
 1 0 1 1 1 0 1 1 0 0 1 0 0 0 0 1 0 0 0 1 0 1 0 1 1 0 1 0 0 0 0 1 1 0 0 1 1
 1 0 1 1 1 1 1 0 0 1 1 0 1 1 0 0 1 0 1 1 1 1 0 1 1 1 1 1 0 1 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 1 1 1 1 1 1 0 1 0 1 1 0 1 1 0 1 0 0 1 1 1 1 1 1 1 1 

In [5]:
# Printing shape of features and target.
print(X.shape)
print(Y.shape)  


(569, 30)
(569,)


##### NOTE: Read why (569,) -> Why the comma?
#### Ans: The comma helps to convey the message that it is a 1-D tupple and not just a number. So it will follow on-tuple operations and not normal numerical mathematical operations.
<a> Source: https://stackoverflow.com/questions/46134891/why-an-extra-comma-in-the-shape-of-a-single-index-numpy-array</a>

In [6]:
# Getting data into dataframe
import pandas as pd
data = pd.DataFrame(cancer_ds.data, columns=cancer_ds.feature_names)
# The column=feature_names, to name the columns as features
data

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0          17.99         10.38          122.80     1001.0          0.11840   
1          20.57         17.77          132.90     1326.0          0.08474   
2          19.69         21.25          130.00     1203.0          0.10960   
3          11.42         20.38           77.58      386.1          0.14250   
4          20.29         14.34          135.10     1297.0          0.10030   
..           ...           ...             ...        ...              ...   
564        21.56         22.39          142.00     1479.0          0.11100   
565        20.13         28.25          131.20     1261.0          0.09780   
566        16.60         28.08          108.30      858.1          0.08455   
567        20.60         29.33          140.10     1265.0          0.11780   
568         7.76         24.54           47.92      181.0          0.05263   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
0             0.27760         0.30010              0.14710         0.2419   
1             0.07864         0.08690              0.07017         0.1812   
2             0.15990         0.19740              0.12790         0.2069   
3             0.28390         0.24140              0.10520         0.2597   
4             0.13280         0.19800              0.10430         0.1809   
..                ...             ...                  ...            ...   
564           0.11590         0.24390              0.13890         0.1726   
565           0.10340         0.14400              0.09791         0.1752   
566           0.10230         0.09251              0.05302         0.1590   
567           0.27700         0.35140              0.15200         0.2397   
568           0.04362         0.00000              0.00000         0.1587   

     mean fractal dimension  ...  worst radius  worst texture  \
0                   0.07871  ...        25.380          17.33   
1                   0.05667  ...        24.990          23.41   
2                   0.05999  ...        23.570          25.53   
3                   0.09744  ...        14.910          26.50   
4                   0.05883  ...        22.540          16.67   
..                      ...  ...           ...            ...   
564                 0.05623  ...        25.450          26.40   
565                 0.05533  ...        23.690          38.25   
566                 0.05648  ...        18.980          34.12   
567                 0.07016  ...        25.740          39.42   
568                 0.05884  ...         9.456          30.37   

     worst perimeter  worst area  worst smoothness  worst compactness  \
0             184.60      2019.0           0.16220            0.66560   
1             158.80      1956.0           0.12380            0.18660   
2             152.50      1709.0           0.14440            0.42450   
3              98.87       567.7           0.20980            0.86630   
4             152.20      1575.0           0.13740            0.20500   
..               ...         ...               ...                ...   
564           166.10      2027.0           0.14100            0.21130   
565           155.00      1731.0           0.11660            0.19220   
566           126.70      1124.0           0.11390            0.30940   
567           184.60      1821.0           0.16500            0.86810   
568            59.16       268.6           0.08996            0.06444   

     worst concavity  worst concave points  worst symmetry  \
0             0.7119                0.2654          0.4601   
1             0.2416                0.1860          0.2750   
2             0.4504                0.2430          0.3613   
3             0.6869                0.2575          0.6638   
4             0.4000                0.1625          0.2364   
..               ...                   ...             ...   
564           0.4107                0.2216         

In [7]:
# Appending a column called class (which is the target {0,1})
data['class'] = cancer_ds.target

In [8]:
# The describe() method is used for calculating some statistical data like percentile, mean and std 
# of the numerical values of the Series or DataFrame.

data.describe()

mean radius  mean texture  mean perimeter    mean area  \
count   569.000000    569.000000      569.000000   569.000000   
mean     14.127292     19.289649       91.969033   654.889104   
std       3.524049      4.301036       24.298981   351.914129   
min       6.981000      9.710000       43.790000   143.500000   
25%      11.700000     16.170000       75.170000   420.300000   
50%      13.370000     18.840000       86.240000   551.100000   
75%      15.780000     21.800000      104.100000   782.700000   
max      28.110000     39.280000      188.500000  2501.000000   

       mean smoothness  mean compactness  mean concavity  mean concave points  \
count       569.000000        569.000000      569.000000           569.000000   
mean          0.096360          0.104341        0.088799             0.048919   
std           0.014064          0.052813        0.079720             0.038803   
min           0.052630          0.019380        0.000000             0.000000   
25%           0.086370          0.064920        0.029560             0.020310   
50%           0.095870          0.092630        0.061540             0.033500   
75%           0.105300          0.130400        0.130700             0.074000   
max           0.163400          0.345400        0.426800             0.201200   

       mean symmetry  mean fractal dimension  ...  worst texture  \
count     569.000000              569.000000  ...     569.000000   
mean        0.181162                0.062798  ...      25.677223   
std         0.027414                0.007060  ...       6.146258   
min         0.106000                0.049960  ...      12.020000   
25%         0.161900                0.057700  ...      21.080000   
50%         0.179200                0.061540  ...      25.410000   
75%         0.195700                0.066120  ...      29.720000   
max         0.304000                0.097440  ...      49.540000   

       worst perimeter   worst area  worst smoothness  worst compactness  \
count       569.000000   569.000000        569.000000         569.000000   
mean        107.261213   880.583128          0.132369           0.254265   
std          33.602542   569.356993          0.022832           0.157336   
min          50.410000   185.200000          0.071170           0.027290   
25%          84.110000   515.300000          0.116600           0.147200   
50%          97.660000   686.500000          0.131300           0.211900   
75%         125.400000  1084.000000          0.146000           0.339100   
max         251.200000  4254.000000          0.222600           1.058000   

       worst concavity  worst concave points  worst symmetry  \
count       569.000000            569.000000      569.000000   
mean          0.272188              0.114606        0.290076   
std           0.208624              0.065732        0.061867   
min           0.000000              0.000000        0.156500   
25%           0.114500              0.064930        0.250400   
50%           0.226700              0.099930        0.282200   
75%           0.382900              0.161400        0.317900   
max           1.252000              0.291000        0.663800   

       worst fractal dimension       class  
count               569.000000  569.000000  
mean                  0.083946    0.627417  
std                   0.018061    0.483918  
min                   0.055040    0.000000  
25%                   0.071460    0.000000  
50%                   0.080040    1.000000  
75%                   0.092080    1.000000  
max                   0.207500    1.000000  

[8 rows x 31 columns]

In [9]:
#Finding how many malignant and benign cancer samples
print(data['class'].value_counts())

1    357
0    212
Name: class, dtype: int64


In [10]:
print(cancer_ds.target_names)

['malignant' 'benign']


In [11]:
data.groupby('class').mean()

mean radius  mean texture  mean perimeter   mean area  mean smoothness  \
class                                                                           
0        17.462830     21.604906      115.365377  978.376415         0.102898   
1        12.146524     17.914762       78.075406  462.790196         0.092478   

       mean compactness  mean concavity  mean concave points  mean symmetry  \
class                                                                         
0              0.145188        0.160775             0.087990       0.192909   
1              0.080085        0.046058             0.025717       0.174186   

       mean fractal dimension  ...  worst radius  worst texture  \
class                          ...                                
0                    0.062680  ...     21.134811      29.318208   
1                    0.062867  ...     13.379801      23.515070   

       worst perimeter   worst area  worst smoothness  worst compactness  \
class                                                                      
0           141.370330  1422.286321          0.144845           0.374824   
1            87.005938   558.899440          0.124959           0.182673   

       worst concavity  worst concave points  worst symmetry  \
class                                                          
0             0.450606              0.182237        0.323468   
1             0.166238              0.074444        0.270246   

       worst fractal dimension  
class                           
0                     0.091530  
1                     0.079442  

[2 rows x 30 columns]

In [12]:
# Train-Test Split
from sklearn.model_selection import train_test_split

In [13]:
X = data.drop('class', axis=1)
Y = data['class']

In [14]:
# Data-type of X: Pandas Dataframe
type(X)

pandas.core.frame.DataFrame

In [15]:
# Splitting the training and testing data by first shuffling it and giving 10% of data as the test dataset.
# Can also do cross-fold validation, by dividing test and train set into various folds.

X_train, X_test, Y_train, Y_test = train_test_split(X,Y, test_size=0.1, random_state=10)
print(X_train.shape, Y_train.shape)
print(X_test.shape, Y_test.shape)

(512, 30) (512,)
(57, 30) (57,)


In [16]:
print(Y.shape, Y_train.shape, Y_test.shape)

(569,) (512,) (57,)


In [17]:
# Printing means
print("X-Train means-")
print(X_train.mean())
print()
print("X-Test means-")
print(X_test.mean())
print()
print("X means-")
print(X.mean())

X-Train means-
mean radius                 14.182154
mean texture                19.248730
mean perimeter              92.356582
mean area                  660.510937
mean smoothness              0.096534
mean compactness             0.104850
mean concavity               0.088763
mean concave points          0.049375
mean symmetry                0.181020
mean fractal dimension       0.062794
radius error                 0.409356
texture error                1.219511
perimeter error              2.906516
area error                  41.060777
smoothness error             0.007024
compactness error            0.025488
concavity error              0.031048
concave points error         0.011778
symmetry error               0.020494
fractal dimension error      0.003766
worst radius                16.328064
worst texture               25.613516
worst perimeter            107.717930
worst area                 887.470898
worst smoothness             0.132191
worst compactness            0.2536

In [18]:
mean_radius_max =  X_train['mean radius'].max()
mean_radius_min =  X_train['mean radius'].min()
mean_radius_half =  (X_train['mean radius'].max() +  X_train['mean radius'].min()) / 2.0
print(f'Mean Radius Max: {mean_radius_max}')
print(f'Mean Radius Min: {mean_radius_min}')
print(f'Mean Radius Half: {mean_radius_half}')

Mean Radius Max: 28.11
Mean Radius Min: 6.981
Mean Radius Half: 17.5455


In [19]:
# Pandas.cut() -> Used cut when you need to segment and sort data values into bins/categories. 
# This function is useful for going from a continuous variable to a categorical variable. 

X_train_binarised = X_train.apply(pd.cut, bins=2, labels=[0,1])

**NOTE:** Here the bins=2, meaning we are cutting the X_train into 2 categories. So example, if we looks at "Mean Radius", where the half is 17.5455. So the mean radius < 17.5455, would be labelled as 0 and the other half > 17.5455 would be labelled as 1. (0,1 comes from the label parameter.) 

In [20]:
X_train.head()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
384        13.28         13.72           85.79      541.8          0.08363   
543        13.21         28.06           84.88      538.4          0.08671   
443        10.57         18.32           66.82      340.9          0.08142   
368        21.71         17.25          140.90     1546.0          0.09384   
325        12.67         17.30           81.25      489.9          0.10280   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
384           0.08575         0.05077              0.02864         0.1617   
543           0.06877         0.02987              0.03275         0.1628   
443           0.04462         0.01993              0.01111         0.2372   
368           0.08562         0.11680              0.08465         0.1717   
325           0.07664         0.03193              0.02107         0.1707   

     mean fractal dimension  ...  worst radius  worst texture  \
384                 0.05594  ...         14.24          17.37   
543                 0.05781  ...         14.37          37.17   
443                 0.05768  ...         10.94          23.31   
368                 0.05054  ...         30.75          26.44   
325                 0.05984  ...         13.71          21.10   

     worst perimeter  worst area  worst smoothness  worst compactness  \
384            96.59       623.7           0.11660            0.26850   
543            92.48       629.6           0.10720            0.13810   
443            69.35       366.3           0.09794            0.06542   
368           199.50      3143.0           0.13630            0.16280   
325            88.70       574.4           0.13840            0.12120   

     worst concavity  worst concave points  worst symmetry  \
384          0.28660               0.09173          0.2736   
543          0.10620               0.07958          0.2473   
443          0.03986               0.02222          0.2699   
368          0.28610               0.18200          0.2510   
325          0.10200               0.05602          0.2688   

     worst fractal dimension  
384                  0.07320  
543                  0.06443  
443                  0.06736  
368                  0.06494  
325                  0.06888  

[5 rows x 30 columns]

In [21]:
X_train_binarised.head()

mean radius mean texture mean perimeter mean area mean smoothness  \
384           0            0              0         0               0   
543           0            1              0         0               0   
443           0            0              0         0               0   
368           1            0              1         1               0   
325           0            0              0         0               0   

    mean compactness mean concavity mean concave points mean symmetry  \
384                0              0                   0             0   
543                0              0                   0             0   
443                0              0                   0             1   
368                0              0                   0             0   
325                0              0                   0             0   

    mean fractal dimension  ... worst radius worst texture worst perimeter  \
384                      0  ...            0             0               0   
543                      0  ...            0             1               0   
443                      0  ...            0             0               0   
368                      0  ...            1             0               1   
325                      0  ...            0             0               0   

    worst area worst smoothness worst compactness worst concavity  \
384          0                0                 0               0   
543          0                0                 0               0   
443          0                0                 0               0   
368          1                0                 0               0   
325          0                0                 0               0   

    worst concave points worst symmetry worst fractal dimension  
384                    0              0                       0  
543                    0              0                       0  
443                    0              0                       0  
368                    1              0                       0  
325                    0              0                       0  

[5 rows x 30 columns]

In [22]:
X_test_binarised = X_test.apply(pd.cut, bins=2, labels=[1,0])
X_test_binarised.head()

mean radius mean texture mean perimeter mean area mean smoothness  \
172           1            1              1         1               0   
553           1            0              1         1               1   
374           1            1              1         1               1   
370           0            0              0         1               1   
419           1            0              1         1               0   

    mean compactness mean concavity mean concave points mean symmetry  \
172                0              1                   0             0   
553                1              1                   1             1   
374                1              1                   1             1   
370                0              1                   0             0   
419                1              1                   1             1   

    mean fractal dimension  ... worst radius worst texture worst perimeter  \
172                      1  ...            1             1               1   
553                      1  ...            1             1               1   
374                      1  ...            1             1               1   
370                      1  ...            1             0               1   
419                      1  ...            1             0               1   

    worst area worst smoothness worst compactness worst concavity  \
172          1                0                 1               1   
553          1                1                 1               1   
374          1                1                 1               1   
370          1                1                 1               0   
419          1                1                 1               1   

    worst concave points worst symmetry worst fractal dimension  
172                    0              1                       1  
553                    1              1                       1  
374                    1              0                       1  
370                    0              0                       1  
419                    1              1                       1  

[5 rows x 30 columns]

In [23]:
# Data-type of X_train_binarised -> pandas dataframe
type(X_train_binarised)

pandas.core.frame.DataFrame

In [24]:
X_train_binarised = X_train_binarised.values
X_test_binarised = X_test_binarised.values

In [25]:
type(X_test_binarised)

numpy.ndarray

In [26]:
# ML Model
from random import randint

# NOTE: Here the model tells if the cancer is malignant or benign depending on the number of features contributing towards the prediction.
# Remember num of features as we have converted everything to be categorical.
# VERY INACURATE way of prediction :(
# b is the thresh-hold of the number of features to account for if the cancer is malignant or benign.
# b would be in the range 0 to 30 (max no. of features.)
b = 10 

i = randint(0, X_train.shape[0])
print(i)
print(X_train_binarised[i])
if (np.sum(X_train_binarised[i,:])>=b):
    print("ML Model says its Malignant.")
else:
    print("ML Model says its Benign.")
    
if (Y_train[i] == 1):
    print("Actual Output is Malignant.")
else:
    print("Actual Output is Benign.")

455
[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
ML Model says its Benign.
Actual Output is Malignant.


## Checking accuracy for different values of b.

In [27]:
print()

In [28]:
def get_accuracy(b):
    count = 0
    correct_count = 0
    for i in range(len(X_train_binarised)):
        if (np.sum(X_train_binarised[i,:])>=b):
            pred_output=1
        else:
            pred_output=0

        if (Y_train.iloc[i] == 1):
            actual_output=1
        else:
            actual_output=0
            
        if (pred_output==actual_output):
            correct_count+=1
        count+=1
    return (correct_count*100/count)

In [30]:
for b in range(32):
    acc = get_accuracy(b)
    print(f'Value of b: {b} \t Accuracy: {acc}')

Value of b: 0 	 Accuracy: 62.6953125
Value of b: 1 	 Accuracy: 25.78125
Value of b: 2 	 Accuracy: 17.578125
Value of b: 3 	 Accuracy: 14.84375
Value of b: 4 	 Accuracy: 15.234375
Value of b: 5 	 Accuracy: 16.40625
Value of b: 6 	 Accuracy: 18.5546875
Value of b: 7 	 Accuracy: 21.6796875
Value of b: 8 	 Accuracy: 26.171875
Value of b: 9 	 Accuracy: 28.125
Value of b: 10 	 Accuracy: 29.296875
Value of b: 11 	 Accuracy: 32.2265625
Value of b: 12 	 Accuracy: 32.8125
Value of b: 13 	 Accuracy: 33.0078125
Value of b: 14 	 Accuracy: 34.1796875
Value of b: 15 	 Accuracy: 34.9609375
Value of b: 16 	 Accuracy: 35.9375
Value of b: 17 	 Accuracy: 36.328125
Value of b: 18 	 Accuracy: 36.5234375
Value of b: 19 	 Accuracy: 36.9140625
Value of b: 20 	 Accuracy: 37.109375
Value of b: 21 	 Accuracy: 37.3046875
Value of b: 22 	 Accuracy: 37.3046875
Value of b: 23 	 Accuracy: 37.3046875
Value of b: 24 	 Accuracy: 37.3046875
Value of b: 25 	 Accuracy: 37.3046875
Value of b: 26 	 Accuracy: 37.3046875
Value 